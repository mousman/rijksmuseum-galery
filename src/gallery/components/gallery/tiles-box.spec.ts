import { flushPromises, shallowMount } from '@vue/test-utils'

import TilesBox from './tiles-box.vue'
import Tile from './tile.vue'
import LoadMoreTile from './load-more-tile.vue'
import Placeholder from './placeholder.vue'
import collectionResponse from '@/gallery/composables/__mocks__/collection.json'
import { useRoute } from 'vue-router'

vi.stubEnv(`VITE_API_KEY`, `myKey`)

const searchParams = {
  q: '',
  key: 'myKey',
  imgonly: 'true',
  s: 'relevance',
  ps: '20',
  p: '1',
}

vi.mock(`vue-router`, () => ({
  useRoute: vi.fn(() => ({ params: {}, query: { q: '' }, meta: {} })),
}))

describe('TilesBox', () => {
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any*/
  it('renders with placeholder if no request made', async (context: any) => {
    context.mockGet('https://www.rijksmuseum.nl/api/en/collection/', {
      data: collectionResponse,
      searchParams,
    })

    const wrapper = shallowMount(TilesBox)
    await flushPromises()
    expect(wrapper.findComponent(Placeholder).exists()).toBe(true)
  })

  /* eslint-disable-next-line @typescript-eslint/no-explicit-any*/
  it('renders with items', async (context: any) => {
    context.mockGet('https://www.rijksmuseum.nl/api/en/collection/', {
      data: collectionResponse,
      searchParams: { ...searchParams, q: 'painting' },
    })

    // @ts-expect-error  it's ok
    useRoute.mockImplementationOnce(() => ({
      params: {},
      query: { q: 'painting' },
      meta: {},
    }))

    const wrapper = shallowMount(TilesBox)

    const loading = wrapper.find('.gallery__loading')
    expect(loading.exists()).toBe(true)
    expect(loading.text()).toBe('Loading...')
    const loadingIcon = wrapper.find('.g-loading__icon')
    expect(loadingIcon.exists()).toBe(true)

    await flushPromises()
    expect(wrapper.findAllComponents(Tile)).toHaveLength(collectionResponse.artObjects.length)
    expect(wrapper.findComponent(Tile).props()).toMatchSnapshot()
    expect(wrapper.findComponent(LoadMoreTile)?.exists()).toBe(true)
  })

  /* eslint-disable-next-line @typescript-eslint/no-explicit-any*/
  it('renders an error info', async (context: any) => {
    context.mockGet('https://www.rijksmuseum.nl/api/en/collection/', {
      data: collectionResponse,
      searchParams: { ...searchParams, q: 'painting' },
      statusCode: 403,
    })

    // @ts-expect-error  it's ok
    useRoute.mockImplementationOnce(() => ({
      params: {},
      query: { q: 'painting' },
      meta: {},
    }))

    const wrapper = shallowMount(TilesBox)
    await flushPromises()

    expect(wrapper.find('.gallery__error').exists()).toBe(true)
    expect(wrapper.find('.gallery__error').text()).toBe('It looks like something went wrong !')
  })
})
