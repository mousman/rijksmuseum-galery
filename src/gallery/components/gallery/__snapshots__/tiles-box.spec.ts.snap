// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TilesBox > renders 1`] = `
{
  "artObject": {
    "hasImage": true,
    "headerImage": {
      "guid": "b06a3793-1e02-42b5-beea-a485a29e7834",
      "height": 430,
      "offsetPercentageX": 0,
      "offsetPercentageY": 0,
      "url": "https://lh3.ggpht.com/VhJW8lei0ed6V3alJxuZPx1eU6wqz6lJBs-4s-XBfuH98AyCsGRr9eQTyEoyRCwsFhAEOyOCmE-tBbIAO3tkXUL9iufm=s0",
      "width": 1791,
    },
    "id": "en-SK-C-1672",
    "links": {
      "self": "http://www.rijksmuseum.nl/api/en/collection/SK-C-1672",
      "web": "http://www.rijksmuseum.nl/en/collection/SK-C-1672",
    },
    "longTitle": "Portrait of Alida Christina Assink, Jan Adam Kruseman, 1833",
    "objectNumber": "SK-C-1672",
    "permitDownload": true,
    "principalOrFirstMaker": "Jan Adam Kruseman",
    "productionPlaces": [],
    "showImage": true,
    "title": "Portrait of Alida Christina Assink",
    "webImage": {
      "guid": "a754c3dd-1734-462d-ab2c-c80a32f1a983",
      "height": 7347,
      "offsetPercentageX": 0,
      "offsetPercentageY": 0,
      "url": "https://lh5.ggpht.com/P5o0tug6fqYu5WAV1L8wObu37HGOGrwgRuGLhkx9Ban89a6PwwUOA2uTSEllOUN_S33pBoJi7NoFcMsw89guz14O8TY=s0",
      "width": 5278,
    },
  },
}
`;

exports[`TilesBox > renders with items 1`] = `
{
  "artObject": {
    "hasImage": true,
    "headerImage": {
      "guid": "b06a3793-1e02-42b5-beea-a485a29e7834",
      "height": 430,
      "offsetPercentageX": 0,
      "offsetPercentageY": 0,
      "url": "https://lh3.ggpht.com/VhJW8lei0ed6V3alJxuZPx1eU6wqz6lJBs-4s-XBfuH98AyCsGRr9eQTyEoyRCwsFhAEOyOCmE-tBbIAO3tkXUL9iufm=s0",
      "width": 1791,
    },
    "id": "en-SK-C-1672",
    "links": {
      "self": "http://www.rijksmuseum.nl/api/en/collection/SK-C-1672",
      "web": "http://www.rijksmuseum.nl/en/collection/SK-C-1672",
    },
    "longTitle": "Portrait of Alida Christina Assink, Jan Adam Kruseman, 1833",
    "objectNumber": "SK-C-1672",
    "permitDownload": true,
    "principalOrFirstMaker": "Jan Adam Kruseman",
    "productionPlaces": [],
    "showImage": true,
    "title": "Portrait of Alida Christina Assink",
    "webImage": {
      "guid": "a754c3dd-1734-462d-ab2c-c80a32f1a983",
      "height": 7347,
      "offsetPercentageX": 0,
      "offsetPercentageY": 0,
      "url": "https://lh5.ggpht.com/P5o0tug6fqYu5WAV1L8wObu37HGOGrwgRuGLhkx9Ban89a6PwwUOA2uTSEllOUN_S33pBoJi7NoFcMsw89guz14O8TY=s0",
      "width": 5278,
    },
  },
}
`;

exports[`TilesBox > renders with placeholder if no request made 1`] = `
{
  "artObject": {
    "hasImage": true,
    "headerImage": {
      "guid": "b06a3793-1e02-42b5-beea-a485a29e7834",
      "height": 430,
      "offsetPercentageX": 0,
      "offsetPercentageY": 0,
      "url": "https://lh3.ggpht.com/VhJW8lei0ed6V3alJxuZPx1eU6wqz6lJBs-4s-XBfuH98AyCsGRr9eQTyEoyRCwsFhAEOyOCmE-tBbIAO3tkXUL9iufm=s0",
      "width": 1791,
    },
    "id": "en-SK-C-1672",
    "links": {
      "self": "http://www.rijksmuseum.nl/api/en/collection/SK-C-1672",
      "web": "http://www.rijksmuseum.nl/en/collection/SK-C-1672",
    },
    "longTitle": "Portrait of Alida Christina Assink, Jan Adam Kruseman, 1833",
    "objectNumber": "SK-C-1672",
    "permitDownload": true,
    "principalOrFirstMaker": "Jan Adam Kruseman",
    "productionPlaces": [],
    "showImage": true,
    "title": "Portrait of Alida Christina Assink",
    "webImage": {
      "guid": "a754c3dd-1734-462d-ab2c-c80a32f1a983",
      "height": 7347,
      "offsetPercentageX": 0,
      "offsetPercentageY": 0,
      "url": "https://lh5.ggpht.com/P5o0tug6fqYu5WAV1L8wObu37HGOGrwgRuGLhkx9Ban89a6PwwUOA2uTSEllOUN_S33pBoJi7NoFcMsw89guz14O8TY=s0",
      "width": 5278,
    },
  },
}
`;
